### PR TITLE
Fix New-Item to create symlink to relative path target

### LINKED ISSUE #9127: New-Item -Type SymbolicLink doesn't define symlinks with relative directory targets as directory symlinks
PR  #8783 addressed #3500, which brought the ability to define symlinks whose targets are defined as _relative_ paths.

On Unix-like platforms, symlinks aren't _typed_; the symlink behaves as if it were the same type (file or directory) of whatever type its _target_ is.

On Windows, by contrast, you must explicitly match the target type at symlink creation time for symlinks to _directories_ to act like directories.
Since `New-Item` currently does _not_ allow you to specify a symlink type - enum value `SymbolicLink` is currently the only option - it must itself _infer_ the appropriate type based on the specified _target_.

As an aside: Since this inference requires that the target already _exist_, creating symlinks to non-existing targets - which is a legitimate use case - presents a conceptual problem that must be solved separately - see #9067

Currently, the inference doesn't work if the target is specified as a _relative_ path to an existing target _directory_.

# Steps to reproduce (on Windows)

Note: If you don't have developer mode enabled, run the tests as admin (elevated).

```powershell
Describe "New-Item: symlink with relative path test" {
  BeforeAll {
    Push-Location TestDrive:/
    # Create a directory...
    $null = New-Item -Type Directory someDir
    # ... and a symlink to that directory, as a sibling, using a *relative* path.
    New-Item -Type SymbolicLink someDirLinkRelative -Target someDir
    # ... and a symlink to that directory, as a sibling, using a *relative* path.
    New-Item -Type SymbolicLink someDirLinkAbsolute -Target (Convert-Path someDir)
    # Echo the symlink target information.
    Get-Item someDirLinkRelative, someDirLinkAbsolute | Select LinkType, Target | Out-String | write-verbose -vb
  }
  AfterAll {
    Pop-Location
  }
  
  It "Symlink with absolute path to existing directory behaves like a directory." {
    (Get-Item someDirLinkAbsolute) | Should -BeOfType System.IO.DirectoryInfo
  }

  It "Symlink with relative path to existing directory behaves like a directory." {
    (Get-Item someDirLinkRelative) | Should -BeOfType System.IO.DirectoryInfo
  }
  
}
```

# Expected behavior

Both tests should succeed.

# Actual behavior

The test with the relative target path fails.
That is, a _file_ symlink rather than a directory symlink was mistakenly created.

# Environment data

```none
PowerShell Core v6.2.0-rc.1 on Microsoft Windows 10 Pro (64-bit; Version 1803, OS Build: 17134.590)
```

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Fix #9127. 

Before the fix we check an item existence relative to _system_ CWD. After the fix we do the check based on current _runspace_ CWD. 

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 1]
```diff
@@ -2361,7 +2361,7 @@ public static string LastWriteTimeString(PSObject instance)
                     // non-existing targets on either Windows or Linux.
```
Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 4 failures in `PowerShell-CI-macos`
Hard link and symbolic link tests.New-Item and hard/symbolic links.New-Item can create a symbolic link to nothing
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.169.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 499
499:             Test-Path $symLinkToNothing | Should -BeTrue
```
Hard link and symbolic link tests.New-Item and hard/symbolic links.New-Item emits an error when path to symbolic link already exists.
```powershell
Expected an exception, with FullyQualifiedErrorId 'SymLinkExists,Microsoft.PowerShell.Commands.NewItemCommand' to be thrown, but the FullyQualifiedErrorId was 'AccessException,Microsoft.PowerShell.Commands.NewItemCommand'. from /Users/runner/runners/2.169.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1:505 char:15
    + ?           { New-Item -ItemType SymbolicLink -Path $realDir -Value $sy ?
    +               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
at <ScriptBlock>, /Users/runner/runners/2.169.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1: line 505
505:             { New-Item -ItemType SymbolicLink -Path $realDir -Value $symLinkToDir -ErrorAction Stop } | Should -Throw -ErrorId "SymLinkExists,Microsoft.PowerShell.Commands.NewItemCommand"
```
New-Item with links.Should create a symbolic link to a non-existing file without error
```powershell
Expected $true, but got $false.
at <ScriptBlock>, /Users/runner/runners/2.169.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1: line 225
225:         Test-Path $FullyQualifiedLink | Should -BeTrue
```
New-Item with links.New-Item -ItemType SymbolicLink should understand directory path ending with slash
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, /Users/runner/runners/2.169.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1: line 255
255:         $symbolicLinkPath | Should -Not -BeNullOrEmpty
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-windows`
New-Item with links.New-Item -ItemType SymbolicLink should understand directory path ending with slash
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\New-Item.Tests.ps1: line 255
255:         $symbolicLinkPath | Should -Not -BeNullOrEmpty
```

**Comment by vexx32** [comment]
@iSazonov thanks for taking the time to sort this one out. 😊 

Quick question -- should / will this affect the `-ItemType Junction` / `-ItemType Hardlink` modes for `New-Item` and should we be testing those here as well?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 2]
```diff
@@ -2361,12 +2361,23 @@ public static string LastWriteTimeString(PSObject instance)
...
```
Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-windows`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got C:\Users\VssAdministrator\AppData\Local\Temp\d4090637-869c-43f9-a060-9e324846e3c1\someDirLinkRelative with type [System.IO.FileInfo].
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\New-Item.Tests.ps1: line 289
289:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 3]
```diff
@@ -2361,12 +2361,22 @@ public static string LastWriteTimeString(PSObject instance)
...
```
Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got /private/tmp/f4fbc540-01e0-42f5-989c-8d78f82b57c4/someDirLinkRelative with type [System.IO.FileInfo].
at <ScriptBlock>, /Users/runner/runners/2.169.0/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1: line 289
289:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```

**Comment by adityapatwardhan** [comment]
@iSazonov Please have a look at the test failure

**Comment by iSazonov** [comment]
@mklement0 Could you please help with the MacOs fail? Do you have any thoughts?

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs (hunk 3 above)
Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got /private/tmp/6f039e50-dab9-49f0-9a51-58d6ec7feb21/someDirLinkRelative with type [System.IO.FileInfo].
at <ScriptBlock>, /Users/runner/runners/2.169.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1: line 289
289:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got /private/tmp/6990029e-f7fb-4110-a26a-b634f5ba29a5/someDirLinkRelative with type [System.IO.FileInfo].
at <ScriptBlock>, /Users/runner/runners/2.169.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1: line 302
302:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 2 failures in `PowerShell-CI-windows`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got $null with type $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\New-Item.Tests.ps1: line 303
303:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```
New-Item: symlink with absolute/relative path test.Symlink with absolute path to existing file behaves like a file
```powershell
someFileLinkAbsolute
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\New-Item.Tests.ps1: line 307
307:         New-Item -Type SymbolicLink someFileLinkAbsolute -Target (Convert-Path someFile)

```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got /private/tmp/2c640d70-dfef-41a6-9942-f9962d7a3678/someDirLinkRelative with type [System.IO.FileInfo].
at <ScriptBlock>, /Users/runner/runners/2.169.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1: line 304
304:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 2 failures in `PowerShell-CI-windows`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got $null with type $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\New-Item.Tests.ps1: line 304
304:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```
New-Item: symlink with absolute/relative path test.Symlink with absolute path to existing file behaves like a file
```powershell
someFileLinkAbsolute
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\New-Item.Tests.ps1: line 308
308:         New-Item -Type SymbolicLink someFileLinkAbsolute -Target (Convert-Path someFile)

```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got /private/tmp/57a90126-0e88-497b-9913-c9da774cedaf/someDirLinkRelative with type [System.IO.FileInfo].
at <ScriptBlock>, /Users/runner/runners/2.169.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1: line 305
305:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 2 failures in `PowerShell-CI-windows`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got $null with type $null.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\New-Item.Tests.ps1: line 305
305:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```
New-Item: symlink with absolute/relative path test.Symlink with absolute path to existing file behaves like a file
```powershell
someFileLinkAbsolute
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\New-Item.Tests.ps1: line 309
309:         New-Item -Type SymbolicLink someFileLinkAbsolute -Target (Convert-Path someFile)

```

Comment (PoshChan) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:2361]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
New-Item: symlink with absolute/relative path test.Symlink with relative path to existing directory behaves like a directory
```powershell
Expected the value to have type [System.IO.DirectoryInfo] or any of its subtypes, but got /private/tmp/4ceece43-75a8-4f68-b703-fd7b6204d71b/someDirLinkRelative with type [System.IO.FileInfo].
at <ScriptBlock>, /Users/runner/runners/2.169.1/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1: line 304
304:         Get-Item someDirLinkRelative | Should -BeOfType System.IO.DirectoryInfo
```

**Review by adityapatwardhan — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@iSazonov Thank you for your contribution!

**Comment by MatejKafka** [comment]
Is this really how relative path for symlink should behave? This PR changes the behavior of the checks from having `-Target` relative to .NET CWD to PowerShell CWD, but it's still inconsistent with how filesystem actually resolves the symlink.

Relative path in the symlink target is resolved by filesystem relative **to the directory containing the symlink**, not relative to CWD of pwsh instance that created it.

**Comment by iSazonov** [comment]
@MatejKafka Please create new issue if you see an inconsistency.

**Comment by mklement0** [comment]
@MatejKafka, while you are correct, this problem doesn't currently surface, although there are related ones:

With respect to _creating_ the symlink, i.e. what target path is stored in its definition:

* On Windows, the relative path is stored as-is (excepting possible path-separator normalization) in the new symlink - which is indeed what should happen.

* On Unix-like platforms, unexpected resolution to a full path happens before the symlink is created, which is a bug, and was reported in #15233.

With respect to _validating the existence_ of the target path on creation:

* This validation - which should indeed happen relative to the _symlink_'s location -  is currently broken on all platforms,  so that targeting a non-existing path is quietly possible, with the exception of using non-existent drive names / UNC paths.

  * While this is _sometimes_ desired, it should be an _opt-in_, and on Windows presents the additional challenge of then needing to specify whether the hypothetical future target is a _file_ or a _directory_.

  * See https://github.com/PowerShell/PowerShell/issues/9067#issuecomment-470140963

**Comment by MatejKafka** [comment]
@mklement0 Actually, I think it currently surfaces. Because the checks are relative to cwd, they incorrectly detect whether the target is file or directory. Let me try to illustrate it with more examples:

(I'll use D:\test as our testing directory to make it clear when I'm talking about absolute paths)

1)
```
mkdir D:/test
mkdir D:/test/dir
ni -Type SymbolicLink D:/test/dir_link -Target ./dir
```
Expected: directory symlink to `D:/test/dir`, with target `./dir`
Got: **file** symlink to `D:/test/dir`, with target `./dir`
The target type detection code checks the relative path `./dir` (`D:/dir`), which doesn't exist, so the symlink incorrectly defaults to**file symlink**.

2)
```
mkdir D:/test
mkdir D:/test/dir
cd D:/test
ni -Type SymbolicLink D:/test/dir_link -Target ./dir
```
Expected: directory symlink to `D:/test/dir`, with target `./dir`
Got: what I expected
Only difference from previous example is the `cd D:/test`. This correctly creates a directory symlink (because the code checks type of target for `./dir`, which now matches the real target).

3)
```
mkdir D:/test/subdir
mkdir D:/test/item # <-- directory
ni D:/test/subdir/item # <-- file
cd D:/test
ni -Type SymbolicLink D:/test/subdir/item_link -Target ./item
```
Expected: file symlink to `D:/test/subdir/item`, with target `./item`
Got: **directory** symlink `D:/test/subdir/item`, with target `./item`
The checks find the file `D:/test/item`, it's a directory, so a directory symlink is created (but the actual target is `D:/test/subdir/item`, which is a file).

**Comment by mklement0** [comment]
Good point, @MatejKafka. As @iSazonov has advised, please create a new issue with your findings.

**Pull request merged**